File: src/ts/magma/app/compile/rule/TypeRule.ts
import { Node } from "../../../../magma/app/compile/node/Node";
import { Rule } from "../../../../magma/app/compile/rule/Rule";
import { Option } from "../../../../magma/api/option/Option";
export class TypeRule implements Rule<Node> {
	type: string;
	childRule: Rule<Node>;
	constructor (type: string, childRule: Rule<Node>) {
		this.type = type;
		this.childRule = childRule;
	}
	lex(input: string): Option<Node> {
		return childRule.lex(input).map(result -  > result.retype(type))/*unknown*/;
	}
}

File: src/main/java/magma/compile/rule/TypeRule.java
package magma.compile.rule;

import magma.compile.Node;
import magma.compile.context.NodeContext;
import magma.compile.error.CompileError;
import magma.result.Err;
import magma.result.Result;

public record TypeRule(String type, Rule rule) implements Rule {
	@Override
	public Result<Node, CompileError> lex(String content) {
		return rule.lex(content).map(node -> node.retype(type));
	}

	@Override
	public Result<String, CompileError> generate(Node node) {
		if (node.is(type)) return rule.generate(node);
		else return new Err<>(new CompileError("Type '" + type + "' not present", new NodeContext(node)));
	}
}
